mplc/multi_partner_learning.py

```diff
@@ -663,6 +664,724 @@ def fit_minibatch(self):
+            self.best = self.history[self.monitor_value][-1]
+        else:
+            self.wait += 1
+            if self.wait >= self.patience:
+                return True
+        return False
+
+    def fit(self):
+
+        # TF function definition
+        @tf.function
+        def aggregate_weights(weights, agg_w):
+            res = list()
+            for weights_per_layer in zip(*weights):
+                res.append(tf.tensordot(weights_per_layer, agg_w, [0, 0]))
+            return res
+
+        @tf.function
+        def fit_minibatch(model, partners_minibatches, partners_optimizers, partners_weights, agg_w):
+            for p_id, minibatch in enumerate(partners_minibatches):  # minibatch == (x,y)
+                # minibatch[0] in a tensor of shape=(number of batch, batch size, img).
+                # We cannot iterate on tensors, so we convert this tensor to a list of
+                # *number of batch* tensors with shape=(batch size, img)
+                x_minibatch = tf.unstack(minibatch[0], axis=0)
+                y_minibatch = tf.unstack(minibatch[1], axis=0)  # same here, with labels
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # set the model weights to partner's ones
+                    model_w.assign(partner_w.read_value())
+
+                for x, y in zip(x_minibatch, y_minibatch):  # iterate over batches
+                    with tf.GradientTape() as tape:
+                        y_pred = model(x)
+                        loss = model.compiled_loss(y, y_pred)
+                    model.compiled_metrics.update_state(y, y_pred)  # log the loss and accuracy
+                    partners_optimizers[p_id].minimize(loss, model.trainable_weights,
+                                                       tape=tape)  # perform local optimization
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # update the partner's weights
+                    partner_w.assign(model_w.read_value())
+            # at the end of the minibatch, aggregate all the local weights
+            aggregated_weights = aggregate_weights(tuple(partners_weights),
+                                                   agg_w)
+
+            # inform the partners with the new weights
+            for p_id in range(len(partners_minibatches)):
+                for new_w, partner_w in zip(aggregated_weights, partners_weights[p_id]):
+                    partner_w.assign(new_w.read_value())
+
+        # Execution
+        self.timer = time.time()
+        for e in range(self.epoch_count):
+            self.epoch_timer = time.time()
+            for partners_minibatches in zip(*self.train_dataset):  # <- partners_minibatches == [(x, y)] * nb_partners
+                fit_minibatch(self.model, partners_minibatches, self.partners_optimizers, self.partners_weights,
```

Comment:
Is this a recursive call??

Reply:
No, it's a indentation mistake, thank you ! We define fit_minibatch in fit, and call fit_minibatch in fit.

Reply:
In fact there is not indentation mistake.  It is just that we define the fit_minibatch, then call it